DataOnion/Auth/TwoFactorAuthService.cs

```diff
@@ -0,0 +1,260 @@
+                {
+                    recentRequest = recentCallRequest;
+                }
+            }
+
+            return recentRequest;
+        }
+
+        public async Task DeleteRequestsAsync(
+            string userId
+        )
+        {
+            await _redisContext.DeleteTwoFactorRequestsAsync(
+                userId
+            );
+        }
+
+        // Checks the TwoFactorRequest's timestamp to see if it has passed the throttle timeout yet.
+        private (bool, int) CheckTwoFactorTimeout(TwoFactorRequest? existingRequest, int throttleTimeout)
+        {
+            if (existingRequest == null)
+            {
+                return (false, 0);
+            }
+            
+            var timeoutSpan = TimeSpan.FromSeconds(throttleTimeout);
+            var createdAt = existingRequest.CreatedAt;
+            var now = DateTime.UtcNow;
+
+            // Get the difference between now and when the request was created.
+            var diff = now - createdAt;
+
+            // If the difference between now and when the request was made is greater than
+            // the timeout period then the request is not valid. Calculate 
+            // the seconds remaining by subtracting the timeout from the difference in
+            // seconds between now and when the request was created.
+
+            var ret = (
+                diff > timeoutSpan,
+                throttleTimeout - (int) diff.TotalSeconds
+            );
+
+            return ret;
+        }
+    }
+
+    public class RegisterDidParams
+    {
+        public string UserId { get; private set; }
+        public string Did { get; private set; }
+        public string Language { get; private set; }
+        public VerificationMethod Method { get; private set; }
+        public int VerificationCode { get; private set; }
+        public int ThrottleTimeout { get; private set; }
+
+        public RegisterDidParams(
+            string userId,
+            string did,
+            string language,
+            VerificationMethod method,
+            int verificationCode,
+            int throttleTimeout
+        )
+        {
+            UserId = userId;
+            Did = did;
+            Language = language;
+            Method = method;
+            VerificationCode = verificationCode;
+            ThrottleTimeout = throttleTimeout;
+        }
+    }
+
+    public class FetchRequestParams
+    {
+        public int Code { get; private set; }
+        public string UserId { get; private set; }
+        public VerificationMethod Method { get; private set; }
+
+        public FetchRequestParams(
+            int code,
+            string userId,
+            VerificationMethod method
+        )
+        {
+            Code = code;
+            UserId = userId;
+            Method = method;
+        }
+    }
+
+    public class RegisterDidResult
+    {
+        public bool AreThereTooManyRequests;
+        public TwoFactorRequest? Request;
```

Comment:
If `Request` is never null when `AreThereTooManyRequests` is `false`, I believe you can use the `MemberNotNullWhen` attribute here:
```cs
[MemberNotNullWhen(false, nameof(Request))]
public bool AreThereTooManyRequests { get; private set; }

public TwoFactorRequest? Request { get; private set; }
```

Reply:
Thanks, I wasn't aware of this annotation. Added this.